packages/cfg-core/lib/ControlFlowGraph.ts

```diff
@@ -18,51 +18,84 @@
 
 import { Node, NodeType } from "./nodes/Node";
 import { Edge } from "./Edge";
-import { Pair } from "./Pair";
-import { RootNode } from "./nodes/RootNode";
+import { Pair } from "./util/Pair";
+import { deepCopy } from "./util/DeepCopy";
 
 export class ControlFlowGraph {
-  private _nodes: Node[];
-  private _edges: Edge[];
+  private readonly _nodes: Node[];
+  private readonly _edges: Edge[];
 
-  constructor() {
-    this._nodes = [];
-    this._edges = [];
+  constructor(nodes: Node[], edges: Edge[]) {
+    this._nodes = deepCopy<Node[]>(nodes);
+    this._edges = deepCopy<Edge[]>(edges);
   }
 
   get nodes(): Node[] {
     return this._nodes;
```

Comment:
instead of making _nodes readonly we can also return [...this._nodes] in the getter function such that the internal datastructure is not exposed.

Reply:
In my opinion, `readonly` is better for the overall codebase, as it is descriptive and defines a fact of ControlFlowGraph immutability with the declaration of parameters and not immutability being implied by the way getters are implemented.